### PR TITLE
Feature/new k8s api layer

### PR DESCRIPTION
**Pull request by ycliuhw**
### Checklist

 - [ ] ~Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?~
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

Add a new API layer for k8s provider;
split from https://github.com/juju/juju/pull/11836

### CODE REVIEW

File: caas/kubernetes/provider/resources/applier.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+)
+
+var logger = loggo.GetLogger("juju.kubernetes.provider.resources")
+
+type applier struct {
+	ops []operation
+}
+
+// NewApplier creates a new applier.
+func NewApplier() Applier {
```
Comment (wallyworld) [inline caas/kubernetes/provider/resources/applier.go:21]: I wonder if "Applier" could be ClusterOperation so it is named like our ModelOperation interface

File: caas/kubernetes/provider/resources/inteface.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+//go:generate go run github.com/golang/mock/mockgen -package mocks -destination mocks/resources_mock.go github.com/juju/juju/caas/kubernetes/provider/resources Resource,Applier
+
+const (
+	// JujuFieldManager marks the resource changes were made by Juju.
+	JujuFieldManager = "juju"
+)
+
+// Resource defines methods for manipulating a k8s resource.
+type Resource interface {
```
Comment (wallyworld) [inline caas/kubernetes/provider/resources/inteface.go:20]: My preference is for a service oriented approach where there's entities which are used as args to various service methods (could be varidac for create, delete etc). Entities encapsulate state but not interactions with the cluster.
So there'd be a Resource interface with Clone(), String().
There'd be a Service interface with Apply(), Get(), Delete() - these take one or more entity resources as args.
Then there'd be a DeploymentService, a StatefulsetService, a DaemonsetService etc. These service instances hold a k8s client reference.

k8s itself takes this approach where there's a client that is akin to the service instance and the domain entities are the Deployment or Secret or whatever.

**Review by hpidcock — APPROVED** [review]
LGTM thank-you

**Review by wallyworld — COMMENTED** [review]
Looks good but I have a different view on approach which would be good to at least discuss before landing.

File: caas/kubernetes/provider/resources/inteface.go (hunk 2 above)
Comment (wallyworld) [reply caas/kubernetes/provider/resources/inteface.go:20]: I guess the Applier provides a different level of abstraction / operation grouping which is an alternative approach to take. So maybe my apprehension is misplaced.

Comment (hpidcock) [reply caas/kubernetes/provider/resources/inteface.go:20]: Lets chat about this tomorrow.

Comment (ycliuhw) [reply caas/kubernetes/provider/resources/inteface.go:20]: As discussed, we keep the current approach for now.

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**